### PR TITLE
Add validation of requirements

### LINKED ISSUE #282: No Exception thrown if solverdummy is configured wrongly
You can reproduce the bugs, if you run the [python solverdummy](https://github.com/precice/precice/tree/develop/tools/solverdummies/python) with the following input.

### Bug1: identical solver and mesh name is used

**terminal1:** `python solverdummy.py precice-config.xml SolverOne MeshOne`

**terminal2:** `python solverdummy.py precice-config.xml SolverOne MeshOne`

* what we observe: preCICE hangs
* what we would like to see: preCICE should throw an exception in terminal2, that SolverOne has already been registered in terminal1

### Bug2: wrong combination of SolverTwo and MeshOne

**terminal1:** `python solverdummy.py precice-config.xml SolverTwo MeshOne`

**terminal2:** `python solverdummy.py precice-config.xml SolverOne MeshOne`

* what we observe: preCICE crashes
* what we would like to see: preCICE should already see in terminal1 that the solverdummy is wrongly configured and throw an exception

### Working: wrong combination of SolverOne and MeshTwo

**terminal:** `python solverdummy.py precice-config.xml SolverOne MeshTwo`

* what we observe: preCICE throws an exception `preCICE:ERROR:  Mesh with name "MeshTwo" is not defined!`. Something similar would be nice for **Bug2**

### PR DESCRIPTION
**Pull request by fsimonis**
This PR adds ID validation and requirement checks to the `SolverInterfaceImpl`.
This reduces code repetition, make the code clearer and hardens our interface against misuse.

It thus fixes #282

## Validation:
* `PRECICE_VALIDATE_MESH_ID(meshID)`
* `PRECICE_VALIDATE_DATA_ID(dataID)`

### Mesh 
* `PRECICE_REQUIRE_MESH_USE`
* `PRECICE_REQUIRE_MESH_PROVIDE` _additionally checks whether the mesh is provided_
* `PRECICE_REQUIRE_MESH_MODIFY` _additionally checks the lock of the mesh_

### Data
* `PRECICE_REQUIRE_DATA_READ` 
* `PRECICE_REQUIRE_DATA_WRITE`

## Example Usage:
In `readScalarData(in toDataID, ...)` specify the requirement `PRECICE_REQUIRE_READ_DATA(toDataID)`.

# TODO
* [x] Turn functions into macros to preserve the call location
* [x] Fix breaking tests

### CODE REVIEW

**Comment by fsimonis** [comment]
Breaking Test:
https://github.com/precice/precice/blob/36b515a6c8d21e61a06679e9c21bb442e28d8785/src/precice/tests/SerialTests.cpp#L978-L980

File: src/precice/impl/SolverInterfaceImpl.cpp [hunk 1]
```diff
@@ -757,11 +764,12 @@ int SolverInterfaceImpl:: setMeshEdge
   int secondVertexID )
 {
   TRACE(meshID, firstVertexID, secondVertexID );
+  PRECICE_VALIDATE_MESH_ID(meshID);
   if ( _clientMode ){
     return _requestManager->requestSetMeshEdge ( meshID, firstVertexID, secondVertexID );
   }
   else {
-    CHECK(not _couplingScheme->isInitialized(), "Edges can only be defined before initialize() is called");
```
Comment (uekerman) [inline src/precice/impl/SolverInterfaceImpl.cpp]: These lines have actually nothing to do with this PR, but with #228 
Will result in a merge conflict.

File: src/precice/impl/ValidationMacros.hpp [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+/// Implementation of PRECICE_REQUIRE_MESH_READ
+#define PRECICE_REQUIRE_MESH_READ_IMPL(id) \
+    PRECICE_VALIDATE_MESH_ID_IMPL(id) \
+    CHECK(_accessor->isMeshUsed(id), "Mesh is not used by this participant! ID:" << id);
+
+/// Implementation of PRECICE_REQUIRE_MESH_WRITE
+#define PRECICE_REQUIRE_MESH_WRITE_IMPL(id) \
+    PRECICE_REQUIRE_MESH_READ_IMPL(id) \
+    MeshContext& context = _accessor->meshContext(id); \
+    CHECK(context.provideMesh, "Can only write to provided meshes! " << id);
```
Comment (uekerman) [inline src/precice/impl/ValidationMacros.hpp]: I don't like the "write" here too much. What about: "Mesh elements (vertices, edges, triangles, quads) can only be created for provided meshes." ?

**Comment by uekerman** [comment]
@fsimonis I understand your reasoning why you also use `read` and `write` for meshes. These names, though, have not been used before and might be easily confused with the read and write of data. I would favor `PRECICE_REQUIRE_MESH_USE` and `PRECICE_REQUIRE_MESH_PROVIDED`. Then, it would be consistent with the naming scheme used so far and with the configuration.   
What do you think?

**Comment by uekerman** [comment]
I currently get
```
/home/uekerman/Repos/precice/src/precice/tests/SerialTests.cpp(966): error: in "PreciceTests/Serial/testStationaryMappingWithSolverMesh": check totalForce == expected has failed [5
5
5 != 5
5
5]
```
which I honestly don't understand. Any clue why `dim=2` behaves different than `dim=3`???
Until now, `dim=3` has never been tested, btw :scream: 

https://github.com/precice/precice/blob/36b515a6c8d21e61a06679e9c21bb442e28d8785/src/precice/tests/SerialTests.cpp#L849

**Comment by fsimonis** [comment]
@uekerman Use `precice::math::equals` to compare vectors for _approximate equality_.

**Comment by floli** [comment]
> @uekerman Use `precice::math::equals` to compare vectors for _approximate equality_.

And if comparing in tests, use `testing::equals`.

**Comment by fsimonis** [comment]
@floli Fixed the `equals`

**Comment by fsimonis** [comment]
I added an additional level of requirement `MODIFY`for the mesh which extends `PROVIDE` by checking the lock of the mesh.

**Pull request merged**